chainer/link.py

```diff
@@ -552,35 +646,57 @@ def __getitem__(self, name):
         """Equivalent to getattr."""
         return getattr(self, name)
 
+    def __setattr__(self, name, value):
+        if self.within_init_scope and isinstance(value, Link):
+            if hasattr(self, name):
+                raise AttributeError(
+                    'cannot register a new link %s: attribute exists' % name)
+            value.name = name
+            self._children.add(name)
+        super(Chain, self).__setattr__(name, value)
+
+    def __delattr__(self, name):
+        self._children.discard(name)
+        super(Chain, self).__delattr__(name)
+
     def add_link(self, name, link):
         """Registers a child link to this chain.
 
-        The registered link is saved and loaded on serialization and
-        deserialization, and involved in the optimization. The registered link
-        is called a child. The child link is set to an attribute of the chain
-        with the given name.
+        .. deprecated:: v2.0.0
+
+           Assign the child link directly to an attribute within
+           :meth:`an initialization scope <chainer.Link.init_scope>`, instead.
+           For example, the following code
+
+           .. code-block:: python
+
+              chain.add_link('l1', L.Linear(3, 5))
 
-        This method also sets the :attr:`~Link.name` attribute of the
-        registered link. If the given link already has the name attribute set,
-        then it raises an error.
+           can be replaced by the following line.
+
+           .. code-block:: python
+
+              with self.init_scope():
+                  chain.l1 = L.Linear(3, 5)
+
+           The latter one is easier for IDEs to keep track of the attribute's
+           type.
 
         Args:
             name (str): Name of the child link. This name is also used as the
                 attribute name.
             link (Link): The link object to be registered.
 
         """
-        if link.name is not None:
-            raise ValueError(
-                'given link is already registered to another chain by name %s'
-                % link.name)
```

Comment:
Almost same as the problem in `Link.add_param`, removal of this code changes the behavior of `Link.add_link`.